bitvm/src/bn254/msm.rs

```diff
@@ -1,267 +1,256 @@
+            let entry = (*p_mul.last().unwrap() + doubled_base).into_affine();  // [a_i] (2^(w.i) P)
+            p_mul.push(entry);
         }
-    }    
-
-    // parameters
-    let mut window = 4;
-    if msm_scalars.len() == 1 {
-        window = 7;
-    } else if msm_scalars.len() == 2 {
-        window = 5;
+
+        let p_mul_scr = {dfs_with_constant_mul(0, window as u32 - 1, 0, &p_mul) };
+        all_tables_scr.push(p_mul_scr);
+        all_tables.push(p_mul);
     }
+    (all_tables, all_tables_scr)
+}
 
-    // MSM
-    let mut acc = ark_bn254::G1Affine::zero();
-    let msm_chunks = G1Affine::hinted_scalar_mul_by_constant_g1(
-        msm_scalars.clone(),
-        msm_bases.clone(),
-        window,
-    );
-    let msm_chunk_hints: Vec<Hint> = msm_chunks.iter().flat_map(|f| f.2.clone()).collect();
-    let msm_chunk_scripts: Vec<Script> = msm_chunks.iter().map(|f| f.1.clone()).collect();
-    let msm_chunk_results: Vec<ark_bn254::G1Affine> = msm_chunks.iter().map(|f| f.0).collect();
-    hints.extend_from_slice(&msm_chunk_hints);
-
-    acc = (acc + msm_acc).into_affine();
-
-    // Additions
-    let mut add_scripts = Vec::new();
-    for i in 0..trivial_bases.len() {
-        // check coeffs before using
-        let (add_script, hint) =
-            G1Affine::hinted_check_add(acc, trivial_bases[i]); // outer_coeffs[i - 1].1
-        add_scripts.push(add_script);
-        hints.extend(hint);
-        acc = (acc + trivial_bases[i]).into_affine();
+// This function computes the slice (w-bit segment) of scalar that is used as an index to 
+// the corresponding row of a table.
+// The index of slice of scalar i.e {a_i} and the index of tables (chunks) i.e {(2^2wi P)} match
+// Output is a value and a script to generate that value from a scalar
+fn get_query_for_table_index(scalar: ark_bn254::Fr, window: usize, table_index: usize) -> (u32, Script) {
+
+    pub fn fq_to_bits(fq: ark_ff::BigInt<4>, limb_size: usize) -> Vec<u32> {
+        let mut bits: Vec<bool> = ark_ff::BitIteratorBE::new(fq.as_ref()).collect();
+        bits.reverse();
+        bits.chunks(limb_size)
+            .map(|chunk| {
+                let mut factor = 1;
+                let res = chunk.iter().fold(0, |acc, &x| {
+                    let r = acc + if x { factor } else { 0 };
+                    factor *= 2;
+                    r
+                });
+                res
+            })
+            .collect()
     }
 
-    // Gather scripts
-    let script = script! {
-        for i in 0..msm_chunk_scripts.len() {
-            // G1Acc preimage
-            if i == 0 {
-                {G1Affine::push( ark_bn254::G1Affine::new_unchecked(ark_bn254::Fq::ZERO, ark_bn254::Fq::ZERO))}
+    // Split Scalar into bits and group window size 
+    let chunks = fq_to_bits(scalar.into_bigint(), window); // {a_0, ..,a_N}
+    // Get Scalar slice (w-bit segment) at index position i.e. a_i
+    let elem = chunks[table_index];
+
+    let scr = script!{
```

Comment:
I think here, this shouldnt be 256, it should be the smallest number greater than Fr::N_BITS divisible by WINDOW, so it should be WINDOW*NUM_TABLES.
```suggestion
    let scr = script!{
        // [scalar]
        {Fr::convert_to_le_bits_toaltstack()}
        // [254-bits]
        for _ in Fr::N_BITS..(NUM_TABLES * WINDOW_G1_MSM as u32) {
            {0}
        }
        // [W*NUM_TABLES-bits]
        for _ in 0..Fr::N_BITS {
            OP_FROMALTSTACK
        }
        for i in 0..NUM_TABLES * WINDOW_G1_MSM {
            if i/WINDOW_G1_MSM == (table_index as u32) {
                OP_TOALTSTACK // preserve all bits for the corresponding table-index
            } else {
                OP_DROP
            }
        }
        for _ in 0..WINDOW_G1_MSM {
            OP_FROMALTSTACK
        }
        // w-bit value a_i
    };
```

Reply:
you're right, this is a cleaner solution.

Reply:
made the fixes in the commit:

https://github.com/BitVM/BitVM/pull/234/commits/810d366fd53c9d37e18b40dfd73a5129ee9747c5